### PR TITLE
Make 404 page a signin page

### LINKED ISSUE #11400: [Question] How redirect to login instead of 404 for private repo?
- Gitea version: 1.11.5
- Git version: 1.14.2
- Operating system: linux-amd64
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x ] Yes
  - [ ] No
  - [ ] Not relevant

## Description
There is the possibility to redirect to _login page_ or show a custom alert message instead of the _404 page_ in this two cases when i try to connect directly form a _url_ of a private repo:

- if I'm not logged in
- my user not have permissions for that repo.

For example i create this repo on [try.gitea](https://try.gitea.io/badstorm/bds_private_rep) 

## Screenshots
Schermata 2020-05-13 alle 12 23 41
Schermata 2020-05-13 alle 12 24 08

### PR DESCRIPTION
**Pull request by zeripath**
Make the 404/http.NotFound page display a login form that will redirect back to the original page if the current user is not signed in

Close #11400 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### Screenshots

Screenshot from 2020-05-16 10-18-47

### CODE REVIEW

File: templates/status/404.tmpl [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
+			<div class="ui divider"></div>
+			<br>
+			<p>{{.i18n.Tr "error404" | Safe}}
+			{{if .ShowFooterVersion}}<p>{{.i18n.Tr "admin.config.app_ver"}}: {{AppVer}}</p>{{end}}
+		</div>
+{{else}}
+	<div class="user signin{{if .LinkAccountMode}} icon{{end}}">
+		{{template "user/auth/signin_navbar" .}}
+		<div class="ui container">
+			<p style="margin-top: 100px"><img class="ui centered image" src="{{StaticUrlPrefix}}/img/404.png" alt="404"/></p>
```
Comment (6543) [inline templates/status/404.tmpl]: we should move css out of tmpl ...

Comment (zeripath) [reply templates/status/404.tmpl]: fine.

**Comment by lafriks** [comment]
Looks kind a strange :) Maybe it would be better just to add link to signin page with correct returnurl? Or other option always redirect to login page for all 404 pages if not signed in

**Comment by CirnoT** [comment]
>Or other option always redirect to login page for all 404 pages if not signed in
Might be a problem for assets, we don't want them redirecting to HTML.

**Comment by guillep2k** [comment]
Only thing that's awkward if signing in to another 404. Is there a simple way to redirect to home when the original page doesn't actually exist?

**Comment by lafriks** [comment]
> Might be a problem for assets, we don't want them redirecting to HTML.

That could be done in router/middleware

**Comment by CirnoT** [comment]
>It is problem already in existing code so that is out of scope for this pr

At least now it returns 404, not 302->200

**Comment by zeripath** [comment]
@guillep2k we cannot do that without revealing the existence or not of a page or repository.

The only answer is to add a small phrase above the login form:

```
If you are sure this page exists and signing in may authorize you to see it...
```

**Comment by zeripath** [comment]
(Or making all 404s once logged-in redirect back to `/`)

**Comment by silverwind** [comment]
The way I handle this in client-side routed applications is that the URL stays the same and it just renders the login page. On sucessful login, actual routing is performed. That way, you don't even need to deal with redirects at all.

**Comment by zeripath** [comment]
@silverwind yes, that would be lovely and clean but without completely refactoring routers/user/home.go and somehow bodging in and making OAuth2, OpenID, 2FA, and SSO all work without storing a redirect - I think we're stuck with a storing a redirect.

**Comment by silverwind** [comment]
I don't think this is the right approach. It'd be better to just emit correct status codes, see https://github.com/go-gitea/gitea/issues/11400#issuecomment-628767361 and remove this double-meaning that our 404s currently have.

**Comment by guillep2k** [comment]
> 
> The only answer is to add a small phrase above the login form:
> 
> ```
> If you are sure this page exists and signing in may authorize you to see it...
> ```

That would be a little better. You know I'm all for being verbose when things get confusing. Or maybe we can just display that message along with another link to the login page (one that's inline with the explanation, besides the one at the header).

But yes, it's a pretty awkward situation: you can't, but I can't tell you that you can't, or maybe you _could_, but there's nothing there. All proper semantics go down the drain. Github is the same; I didn't check others.

**Comment by bagasme** [comment]
@zeripath what's the point of sign in inside 404? I think it is more appropriate for 403 (access denied)

**Comment by CirnoT** [comment]
@bagasme We don't show 403 explicitly in order not to leak whether resource exists or not

**Comment by silverwind** [comment]
See https://github.com/go-gitea/gitea/issues/11400 for more discussion on the status codes.

**Comment by mrsdizzie** [comment]
I don't think any of these changes will really satisfy the various requests about this because the cause of confusion and the behavior people want changed is Gitea returning a 404 error for a repo that exists rather than a 401 or something.

Theres already a sign in link on the top of each page which will give the correct return link. I think this looks a bit weird to duplicate the sign in box on the 404 page just to save that click.

Perhaps the current text we show could be slightly adjusted to make it more clear a user isn't signed in:

```
The page you are trying to reach either does not exist or you are not authorized to view it (sign in).
```

"sign in" being a conditional link that shows up if you are signed out.

**Comment by zeripath** [comment]
@mrsdizzie if they click that signin button they lose the redirect back to this page - that makes it very awkward to use. This will redirect you straight back to the resource - so if the resource does exist and the user should have permission to view it - the link will just work.

**Comment by zeripath** [comment]
This **is** actually a significant UI improvement. Allowing the user to attempt to relogin with a redirect straight back to the page means that they can check if it's genuinely that the page is Not Found for them. 

We can go round and round in more circles about why 404 for this is bad, but it's the only way to prevent leaking data. Even if you think that in your circumstances you will always want the real 401 you **will** be wrong at some point.

There is no other way other than extending 404 to mean NotFound (at least for you).

**Comment by zeripath** [comment]
@guillep2k I've added another locale key: `error404_not_signed_in` and updated the screenshot

**Comment by zeripath** [comment]
Let's be clearer how this actually improves things:

* I am a member of a private repository: http://localhost/gitea/administrator/merge-test
* I get sent an email with the link to an issue:  http://localhost/gitea/administrator/merge-test/issues/101
* My gitea session has timedout for whatever reason so when I click that link I get a 404
* Oh damn, is it wrong...? 
  - I don't see the comment saying I may be unauthorized because all I see is the gigantic 404 so I don't proceed any further...
  - Or I see that I have to sign-in... so I click that
* I get redirected to my dashboard... so I have to actually repaste the link in to get to what I was actually looking at...

This actually significantly improves things because: 
1. the gigantic sign-in box is difficult to miss and is a bit weird - so I am much more likely to read what the page says
2. If I do manage to sign-in I will get redirected straight back to the resource I was trying to look at.

**Review by 6543 — APPROVED** [review]

**Comment by mrsdizzie** [comment]
> @mrsdizzie if they click that signin button they lose the redirect back to this page - that makes it very awkward to use. This will redirect you straight back to the resource - so if the resource does exist and the user should have permission to view it - the link will just work.

I don't think that example  above is what actually happens:

(Be logged out)
Visit link to https://gitea.com/mrsdizzie/fake

See 404 error page. 
Sign In link at top is ```https://gitea.com/user/login?redirect_to=%2fmrsdizzie%2ffake``` 

and signing in takes me back to https://gitea.com/mrsdizzie/fake as a logged in user. As far as I know the  existing  sign in link always sets redirect_to to the page you  were on when you clicked it

https://github.com/go-gitea/gitea/blob/0052d788daf0bdc938d6189da880b5113a1113c6/templates/base/head_navbar.tmpl#L134-L136

So all of the examples above would be  redirected to the original link after singing in. Not that this current interface is the best obviously due to complaints/confusion etc...

**Comment by zeripath** [comment]
OK so mea culpa it does keep the url - but the sign-in link loses the query part so let's try this instead:

http://localhost/gitea/administrator/merge-test/issues?page=5&q=&type=all&sort=&state=open&labels=&milestone=0&assignee=0

The query portion will disappear on the current system.

**Comment by guillep2k** [comment]
> 
> 
> OK so mea culpa it does keep the url - but the sign-in link loses the query part so let's try this instead:
> 
> http://localhost/gitea/administrator/merge-test/issues?page=5&q=&type=all&sort=&state=open&labels=&milestone=0&assignee=0
> 
> The query portion will disappear on the current system.

So maybe we should just fix the _Sign In_ link to include the query string? Or in addition to this PR, perhaps?

**Comment by guillep2k** [comment]
I guess something like this is out of the question?

image

**Comment by CirnoT** [comment]
This looks very weird with OpenID switch header above 404 image

**Pull request closed without merging**